openff/interchange/interop/openmm/_topology.py

```diff
@@ -173,6 +174,30 @@ def to_openmm_topology(
                 order=bond_order,
             )
 
+    if has_virtual_sites and not collate:
+        # TODO: This structure is imperfect, no clue how to handle chains here
+        virtual_site_chain = openmm_topology.addChain(atom_chain_id)
+
+        for molecule_index, molecule in enumerate(topology.molecules):
+            virtual_sites_in_this_molecule: list[VirtualSiteKey] = molecule_virtual_site_map[molecule_index]
+
+            # make a new "residue" for each molecule which has virtual sites ...
+            if len(virtual_sites_in_this_molecule) > 0:
+                # ... but give it the name of the parent molecule
+                this_virtual_site_residue = openmm_topology.addResidue(
+                    name=molecule.atom(0).metadata.get("residue_name", "UNK"),
+                    chain=virtual_site_chain,
+                )
+
+            for this_virtual_site in virtual_sites_in_this_molecule:
+                virtual_site_name = this_virtual_site.name
+
+                openmm_topology.addAtom(
+                    virtual_site_name,
+                    virtual_site_element,
+                    this_virtual_site_residue,
+                )
+
```

Comment:
I like Irfan's suggestions of having residue names and numbers match - that way the only thing that varies with the `collate` argument is the atom ordering, not the metadata of the virtual sites. To that end, I'd suggest an implementation along the lines of the following, which just uses the fact that atom indices for parent atoms are kept constant to index into the chain and residue metadata that we just generated:

```suggestion
    if has_virtual_sites and not collate:
        chain = None
        residue = None
        for virtual_site_key in interchange["VirtualSites"].key_map:
            assert isinstance(virtual_site_key, VirtualSiteKey)
            parent_atom_index = virtual_site_key.orientation_atom_indices[0]
            parent_atom = omm_atoms[parent_atom_index]

            if chain is None or chain.id != parent_atom.residue.chain.id:
                chain = openmm_topology.addChain(id=parent_atom.residue.chain.id)

            if residue is None or not (
                residue.name == parent_atom.residue.name
                and residue.id == parent_atom.residue.id
                and residue.insertionCode == parent_atom.residue.insertionCode
                and residue.chain.id == parent_atom.residue.chain.id
            ):
                residue = openmm_topology.addResidue(
                    name=parent_atom.residue.name,
                    chain=chain,
                    id=parent_atom.residue.id,
                    insertionCode=parent_atom.residue.id,
                )

            virtual_site_name = virtual_site_key.name

            openmm_topology.addAtom(
                name=virtual_site_name,
                element=virtual_site_element,
                residue=residue,
            )
```